sourcegraph/learn
# Upgrade to Next.js 12

### What should this PR do?
Resolves [DEVED-233](https://sourcegraph.atlassian.net/browse/DEVED-233) by upgrading our app to Next.js 12. For more info about the release, see the [announcement blog](https://nextjs.org/blog/next-12).

We're not getting a ton of lift out of this release, for a few reasons:
- We are not able to use the Rust compiler, thanks to our use of `styled-components`. There are plans to port popular libraries like `styled-components` and `emotion` "soon", so my guess is that we will reap these benefits before too long. Either way, that's better than rewriting all of our custom styles to `styled-jsx`.
- We do not currently have a use case for middleware, though that could change.
- We are not currently toggling on React 18 features.

So not a ton specific to this release, but we will get Next.js specific linting callouts here, which will be helpful.

### Why are we making this change?
- By staying current, we can take advantage of new capabilities as they become relevant to us.

### What are the acceptance criteria? 
- Everything should work exactly as it does on prod.

### How should this PR be tested?
- Check out the branch, remove `node_modules`, and install deps with `npm ci`.
- Run `npm run lint`, check that it's working
- Check that everything on the app looks/behaves as you would expect.

## Pull request process

**Reviewers**:

1. Test functionality using the criteria above.
2. Offer tips for efficiency, feedback on best practices, and possible alternative approaches and things that may not have been considered.
3. For shorter, "quick" PRs, use your best judgement on #​2.
4. Use a collaborative approach and provide resources and/or context where appropriate.
5. Provide screenshots/grabs where appropriate to show findings during review.

**Reviewees**:

1. Prefer incremental and appropriately-scoped changes.
2. Leave a comment on things you want explicit feedback on.
3. Respond clearly to comments and questions.
